File: scala/src/main/scala/raytracer/Raytracer.scala
package raytracer

object Raytracer {
  type Pos = Vec3
  type Dir = Vec3
  type Color = Vec3
  final val Black: Color = Vec3(0, 0, 0)
  final val White: Color = Vec3(1, 1, 1)

  final case class Ray(origin: Pos, dir: Dir) {

    inline def pointAtParam(t: Double): Pos = origin + dir.scale(t)

    def aabbHit(aabb: AABB, tMin0: Double, tMax0: Double): Boolean = {
      def go(min_ : Double, max_ : Double, origin_ : Double, dir_ : Double, tMin_ : Double, tMax_ : Double): (Double, Double) = {
        val invD = 1.0 / dir_
        val t0 = (min_ - origin_) * invD
        val t1 = (max_ - origin_) * invD
        val t0_ = if (invD < 0) t1 else t0
        val t1_ = if (invD < 0) t0 else t1
        val tMin__ = t0_ max tMin_
        val tMax__ = t1_ min tMax_
        (tMin__, tMax__)
      }
      val (tMin1, tMax1) = go(aabb.min.x, aabb.max.x, origin.x, dir.x, tMin0, tMax0)
      if(tMax1 <= tMin1) false
      else {
        val (tMin2, tMax2) = go(aabb.min.y, aabb.max.y, origin.y, dir.y, tMin1, tMax1)
        if(tMax2 <= tMin2) false
        else {
          val (tMin3, tMax3) = go(aabb.min.z, aabb.max.z, origin.z, dir.z, tMin2, tMax2)
          tMax3 > tMin3
        }
      }
    }
  }

  final case class Hit(t: Double, p: Pos, normal: Dir, color: Color)
  final case class Sphere(pos: Pos, color: Color, radius: Double) {
    private[this] final val radiusSqrd = radius * radius
    private[this] final val scalaFactor = 1.0 / radius

    def aabb: AABB = {
      val rVec = Vec3(radius, radius, radius)
      AABB(pos - rVec, pos + rVec)
    }

    def hit(ray: Ray, tMin: Double, tMax: Double): Option[Hit] = {
      val oc = ray.origin - pos
      val a = ray.dir dot ray.dir
      val b = oc dot ray.dir
      val c = (oc dot oc) - radiusSqrd
      val discriminant = b * b - a * c

      def tryHit(temp: Double) =
        if (temp < tMax && temp > tMin) {
          val pointAtParam = ray.pointAtParam(temp)
          Some(Hit(temp, pointAtParam, (pointAtParam - pos).scale(scalaFactor), color))
        }
        else None

      if (discriminant <= 0) None
      else {
        val sqrtDiscriminant = math.sqrt(discriminant)

        tryHit((-b - sqrtDiscriminant) / a) match {
          case s: Some[Hit] => s
          case None => tryHit((-b + sqrtDiscriminant) / a)
        }
      }
    }
  }

  type Objs = BVH[Sphere]

  def objsHit(objs: Objs, ray: Ray, tMin: Double, tMax: Double): Option[Hit] = objs match {
    case Leaf(_, sphere) => sphere.hit(ray, tMin, tMax)
    case Split(box, left, right) =>
      if(!ray.aabbHit(box, tMin, tMax)) None
      else objsHit(left, ray, tMin, tMax) match {
        case opt@Some(hit1) => objsHit(right, ray, tMin, hit1.t).orElse(opt)
        case None => objsHit(right, ray, tMin, tMax)
      }
  }

  final case class Camera(origin: Pos, llc: Pos, horizontal: Dir, vertical: Dir)

  object Camera {
    def apply(lookFrom: Pos, lookAt: Pos, vUp: Dir, vFov: Double, aspect: Double): Camera = {
      val theta = vFov * math.Pi / 180.0
      val halfHeight = math.tan(theta / 2.0)
      val halfWidth = aspect * halfHeight
      val origin = lookFrom
      val w = (lookFrom - lookAt).normalise
      val u = (vUp cross w).normalise
      val v = w cross u
      Camera( lookFrom
            , origin - u.scale(halfWidth) - v.scale(halfHeight) - w
            , u.scale(2*halfWidth)
            , v.scale(2*halfHeight)
            )
    }
  }

  def getRay(cam: Camera, s: Double, t: Double): Ray =
    Ray( cam.origin
       , cam.llc + cam.horizontal.scale(s) + cam.vertical.scale(t) - cam.origin
       )

  inline def reflect(v: Vec3, n: Vec3): Vec3 =
    v - n.scale(2 * (v dot n))

  inline def scatter(rayDir: Dir, hit: Hit): Option[Ray] = {
    val reflected = reflect(rayDir.normalise, hit.normal)
    if((reflected dot hit.normal) > 0) Some(Ray(hit.p, reflected)) else None
  }

  def rayColor(objs: Objs, ray: Ray, depth: Int): Color = objsHit(objs, ray, 0.001, 1.0 / 0.0) match {
      case Some(hit) => scatter(ray.dir, hit) match {
        case Some(scattered) if depth < 50 => hit.color * rayColor(objs, scattered, depth+1)
        case _ => Black
      }
      case None =>
        val t = 0.5 * (ray.dir.normalise.y + 1.0)
        Vec3.one.scale(1.0-t) + Vec3(0.5, 0.7, 1).scale(t)
  }

  def traceRay(objs: Objs, width: Int, height: Int, camera: Camera, j: Int, i: Int): Color =
    rayColor(objs,
      getRay(camera,
        i.toDouble / width.toDouble,
        j.toDouble / height.toDouble), 0)

  def render(objs: Objs, width: Int, height: Int, camera: Camera): Image = {
    import raytracer.Image.Pixel
    val colorToPixel = (c: Color) => new Pixel((255.99 * c.x).toByte, (255.99 * c.y).toByte, (255.99 * c.z).toByte)
    Image(width, height, (j, i) => colorToPixel(traceRay(objs, width, height, camera, j, i)))
  }
}

File: elixir/lib/raytracer.ex
defmodule Raytracer do
  def x({x, _y, _z}), do: x
  def y({_x, y, _z}), do: y
  def z({_x, _y, z}), do: z

  def scale({x, y, z}, s) do
    {x * s, y * s, z * s}
  end

  def norm(v), do: v |> dot(v) |> :math.sqrt()

  def normalise(v) do
    scale(v, 1.0 / norm(v))
  end

  def add({x1, y1, z1}, {x2, y2, z2}) do
    {x1 + x2, y1 + y2, z1 + z2}
  end

  def sub({x1, y1, z1}, {x2, y2, z2}) do
    {x1 - x2, y1 - y2, z1 - z2}
  end

  def mult({x1, y1, z1}, {x2, y2, z2}) do
    {x1 * x2, y1 * y2, z1 * z2}
  end

  def dot(v1, v2) do
    {x, y, z} = mult(v1, v2)
    x + y + z
  end

  def cross({x1, y1, z1}, {x2, y2, z2}) do
    {y1 * z2 - z1 * y2, z1 * x2 - x1 * z2, x1 * y2 - y1 * x2}
  end

  def reflect(v1, v2) do
    sub(v1, scale(v2, 2.0 * dot(v1, v2)))
  end

  def centre({{x1, y1, z1}, {x2, y2, z2}}) do
    {x1 + 0.5 * (x2 - x1), y1 + 0.5 * (y2 - y1), z1 + 0.5 * (z2 - z1)}
  end

  def hit(
        {{x1, y1, z1}, {x2, y2, z2}},
        {{x_origin, y_origin, z_origin}, {x_dir, y_dir, z_dir}},
        tmin0,
        tmax0
      ) do
    iter = fn min, max, origin, dir, tmin, tmax ->
      if dir == 0.0 do
        {tmin, tmax}
      else
        inv_d = 1.0 / dir
        t0 = (min - origin) * inv_d
        t1 = (max - origin) * inv_d
        {t0, t1} = if inv_d < 0.0, do: {t1, t0}, else: {t0, t1}
        tmin = max(t0, tmin)
        tmax = min(t1, tmax)
        {tmin, tmax}
      end
    end

    {tmin1, tmax1} = iter.(x1, x2, x_origin, x_dir, tmin0, tmax0)

    if tmax1 <= tmin1 do
      false
    else
      {tmin2, tmax2} = iter.(y1, y2, y_origin, y_dir, tmin1, tmax1)

      if tmax2 <= tmin2 do
        false
      else
        {tmin3, tmax3} = iter.(z1, z2, z_origin, z_dir, tmin2, tmax2)
        tmax3 > tmin3
      end
    end
  end

  def hit({:leaf, _, s}, r, tmin, tmax) do
    hit(s, r, tmin, tmax)
  end

  def hit({:split, b, left, right}, r, tmin, tmax) do
    if not hit(b, r, tmin, tmax) do
      nil
    else
      case hit(left, r, tmin, tmax) do
        {t, _, _, _} = h ->
          if h_ = hit(right, r, tmin, t) do
            h_
          else
            h
          end

        nil ->
          hit(right, r, tmin, tmax)
      end
    end
  end

  def hit({pos, colour, radius}, {origin, dir}, tmin, tmax) do
    oc = sub(origin, pos)
    a = dot(dir, dir)
    b = dot(oc, dir)
    c = dot(oc, oc) - radius * radius
    discriminant = b * b - a * c

    helper = fn temp ->
      if temp < tmax && temp > tmin do
        {temp, point_at_param({origin, dir}, temp),
         scale(sub(point_at_param({origin, dir}, temp), pos), 1.0 / radius), colour}
      else
        nil
      end
    end

    if discriminant <= 0.0 do
      nil
    else
      case helper.((-b - :math.sqrt(b * b - a * c)) / a) do
        nil -> helper.((-b + :math.sqrt(b * b - a * c)) / a)
        x -> x
      end
    end
  end

  def enclosing(
        {{x1, y1, z1}, {x2, y2, z2}},
        {{x3, y3, z3}, {x4, y4, z4}}
      ) do
    {{min(x1, x3), min(y1, y3), min(z1, z3)}, {max(x2, x4), max(y2, y4), max(z2, z4)}}
  end

  def aabb({:leaf, aabb, _}), do: aabb
  def aabb({:split, aabb, _, _}), do: aabb

  @black {0.0, 0.0, 0.0}
  @white {1.0, 1.0, 1.0}

  def point_at_param({origin, dir}, t) do
    add(origin, scale(dir, t))
  end

  def scatter({_origin, dir}, {_t, p, normal, colour}) do
    reflected =
      dir
      |> normalise
      |> reflect(normal)

    scattered = {p, reflected}

    if dot(reflected, normal) > 0.0 do
      {scattered, colour}
    else
      nil
    end
  end

  def colour({origin, dir}, objs, depth) do
    case hit(objs, {origin, dir}, 0.001, 1_000_000_000.0) do
      {_, _, _, _} = hit ->
        case scatter({origin, dir}, hit) do
          {scattered, attenuation} ->
            if depth < 50 do
              mult(attenuation, colour(scattered, objs, depth + 1))
            else
              @black
            end

          nil ->
            @black
        end

      nil ->
        unit_dir = normalise(dir)
        t = 0.5 * (y(unit_dir) + 1.0)
        bg = {0.5, 0.7, 1.0}
        add(scale(@white, 1.0 - t), scale(bg, t))
    end
  end

  def to_aabb({pos, _colour, radius}) do
    {sub(pos, {radius, radius, radius}), add(pos, {radius, radius, radius})}
  end

  def new(all_objs) do
    helper = fn
      _helper, _d, _n, [] ->
        raise "no nodes"

      _helper, _d, _n, [x] ->
        {:leaf, to_aabb(x), x}

      helper, d, n, xs ->
        xs =
          MergeSort.sort(xs, n, fn a, b ->
            {a, b} =
              case Integer.mod(d, 3) do
                0 -> {a |> to_aabb() |> centre() |> x(), b |> to_aabb() |> centre() |> x()}
                1 -> {a |> to_aabb() |> centre() |> y(), b |> to_aabb() |> centre() |> y()}
                _ -> {a |> to_aabb() |> centre() |> z(), b |> to_aabb() |> centre() |> z()}
              end

            a <= b
          end)

        {xs_left, xs_right} = Enum.split(xs, round(n / 2))

        [left, right] =
          if n < 100 do
            [helper.(helper, d + 1, n / 2, xs_left), helper.(helper, d + 1, n - n / 2, xs_right)]
          else
            [
              Task.async(fn -> helper.(helper, d + 1, n / 2, xs_left) end),
              Task.async(fn -> helper.(helper, d + 1, n - n / 2, xs_right) end)
            ]
            |> Task.await_many()
          end

        b = left |> aabb() |> enclosing(aabb(right))
        {:split, b, left, right}
    end

    helper.(helper, 0, length(all_objs), all_objs)
  end

  def new_camera(lookfrom, lookat, vup, vfov, aspect) do
    theta = vfov * :math.pi() / 180.0
    half_height = :math.tan(theta / 2.0)
    half_width = aspect * half_height
    w = sub(lookfrom, lookat) |> normalise()
    u = vup |> cross(w) |> normalise()
    v = cross(w, u)

    {lookfrom, lookfrom |> sub(scale(u, half_width)) |> sub(scale(v, half_height)) |> sub(w),
     scale(u, 2.0 * half_width), scale(v, 2.0 * half_height)}
  end

  def ray({origin, llc, horizontal, vertical}, s, t) do
    {origin,
     llc
     |> add(scale(horizontal, s))
     |> add(scale(vertical, t))
     |> sub(origin)}
  end

  def trace_ray(objs, width, height, cam, j, i) do
    u = i / width
    v = j / height

    cam
    |> ray(u, v)
    |> colour(objs, 0)
  end

  def colour_to_pixel({x, y, z}) do
    {floor(x * 255.99), floor(y * 255.99), floor(z * 255.99)}
  end

  def image_to_ppm(dev, {:image, pixels, height, width}) do
    IO.write(dev, ~s"""
    P3
    #{width} #{height}
    255
    """)

    for {r, g, b} <- pixels do
      IO.write(dev, "#{r} #{g} #{b}\n")
    end
  end

  def render(objs, width, height, cam) do
    pixel = fn l ->
      i = Integer.mod(l, width)
      j = height - l / width

      objs
      |> trace_ray(width, height, cam, j, i)
      |> colour_to_pixel()
    end

    pixels =
      0..(height * width)
      |> ParallelStream.map(pixel)
      |> Enum.to_list()

    {:image, pixels, height, width}
  end

  def from_scene(width, height, {cam_look_from, cam_look_at, cam_fov, spheres}) do
    {new(spheres),
     new_camera(cam_look_from, cam_look_at, {0.0, 1.0, 0.0}, cam_fov, width / height)}
  end

  def rgbbox() do
    n = 10
    k = 60

    leftwall =
      Enum.flat_map(0..n, fn y ->
        Enum.map(0..n, fn z ->
          {{-k / 2.0, -k / 2.0 + k / n * y, -k / 2.0 + k / n * z}, {1.0, 0.0, 0.0}, k / (n * 2.0)}
        end)
      end)

    midwall =
      Enum.flat_map(0..n, fn x ->
        Enum.map(0..n, fn y ->
          {{-k / 2.0 + k / n * x, -k / 2.0 + k / n * y, -k / 2.0}, {1.0, 1.0, 0.0}, k / (n * 2.0)}
        end)
      end)

    rightwall =
      Enum.flat_map(0..n, fn y ->
        Enum.map(0..n, fn z ->
          {{k / 2.0, -k / 2.0 + k / n * y, -k / 2.0 + k / n * z}, {0.0, 0.0, 1.0}, k / (n * 2.0)}
        end)
      end)

    bottom =
      Enum.flat_map(0..n, fn x ->
        Enum.map(0..n, fn z ->
          {{-k / 2.0 + k / n * x, -k / 2.0, -k / 2.0 + k / n * z}, {1.0, 1.0, 1.0}, k / (n * 2.0)}
        end)
      end)

    {{0.0, 30.0, 30.0}, {0.0, -1.0, -1.0}, 75.0,
     Enum.concat([leftwall, midwall, rightwall, bottom])}
  end

  def irreg() do
    n = 100
    k = 600

    spheres =
      Enum.flat_map(0..n, fn x ->
        Enum.map(0..n, fn z ->
          {{-k / 2.0 + k / n * x, 0.0, -k / 2.0 + k / n * z}, @white, k / (n * 2.0)}
        end)
      end)

    {{0.0, 12.0, 30.0}, {0.0, 10.0, -1.0}, 75.0, spheres}
  end

  def main(args \\ []) do
    {opts, scene, _} =
      OptionParser.parse(args,
        strict: [file_out: :string, height: :integer, width: :integer]
      )

    scene =
      case scene do
        ["irreg"] -> irreg()
        ["rgbbox"] -> rgbbox()
        [] -> rgbbox()
        _ -> raise "invalid scene"
      end

    width = Keyword.get(opts, :width, 200)
    height = Keyword.get(opts, :height, 200)

    {objs, cam} = from_scene(width, height, scene)

    result = render(objs, width, height, cam)

    dev =
      case Keyword.fetch(opts, :file_out) do
        {:ok, filename} ->
          File.open!(filename, [:write])

        :error ->
          :stdio
      end

    image_to_ppm(dev, result)
  end
end
